src/tools/server-autosave.js

```diff
@@ -0,0 +1,43 @@
+import {
+  setMostRecentSavedContent,
+  saveNotebookToServer,
+  setConnectionStatus
+} from "../actions/actions";
+import { clearAutosave } from "./autosave";
+
+export function notebookChangedSinceSave(state) {
+  const { previouslySavedContent: previous } = state;
+  return previous.jsmd !== state.jsmd || previous.title !== state.title;
+}
+
+export function checkForServerAutosave(store) {
+  store.dispatch(setMostRecentSavedContent());
+  setInterval(async () => {
+    const state = store.getState();
+    const { username: notebookOwner } = state.notebookInfo;
+    const { name: thisUser } = state.userData;
+    let validAutosave = false;
+    if (
+      notebookChangedSinceSave(state) && // has the notebook changed?
+      thisUser !== undefined && // is this a logged-in-user?
+      notebookOwner === thisUser // is this notebook owned by the current user?
+    ) {
+      try {
+        await store.dispatch(saveNotebookToServer(false));
+        validAutosave = true;
+      } catch (err) {
+        // FIXME: come up with a compelling error case
+        store.dispatch(setConnectionStatus("CONNECTION_LOST"));
+        // console.error(Error(err.message));
+      }
+      if (validAutosave) {
+        store.dispatch(setMostRecentSavedContent());
+        // remove previous autosave if we've successfully saved.
+        clearAutosave(state);
+        if (state.notebookInfo.connectionStatus !== "CONNECTION_ACTIVE") {
+          store.dispatch(setConnectionStatus("CONNECTION_ACTIVE"));
+        }
+      }
+    }
+  }, 10000);
```

Comment:
So I think we should set this to 30 seconds. Rationale:

1. It will make the history viewer more manageable
2. It will reduce the size of the database dramatically
3. Even if the user accidentally closes their tab, autosave should restore it if the user re-opens the window.

Also, could you make this a constant and put it at the top of the file? Something like `AUTOSAVE_INTERVAL` would be fine.